core/command_line_parser.h

```diff
@@ -0,0 +1,246 @@
+	// Make the option visible in help.
+	bool _hidden = true;
+	// If true, arguments can be specified without option name.
+	bool _positional = false;
+	// If true, argument always should be provided.
+	bool _required = false;
+	// If true, the option can be specified several times.
+	bool _multitoken = false;
+	// Arguments count required for the option, -1 for all arguments left.
+	int _args_count = 1;
+	// Checker for each argument, nullptr if not check specified.
+	ArgumentChecker *_checker = nullptr;
+
+protected:
+	static void _bind_methods();
+
+public:
+	using CheckFunction = bool (*)(const String &);
+
+	void set_names(const PackedStringArray &p_names);
+	PackedStringArray get_names() const;
+	void set_default_args(const PackedStringArray &p_args);
+	PackedStringArray get_default_args() const;
+	void set_allowed_args(const PackedStringArray &p_args);
+	PackedStringArray get_allowed_args() const;
+	void set_description(const String &p_description);
+	String get_description() const;
+	void set_category(const String &p_category);
+	String get_category() const;
+	void set_arg_text(const String &p_arg_text);
+	String get_arg_text() const;
+	void set_args_count(int p_count);
+	int get_args_count() const;
+	void set_hidden(const bool p_hidden);
+	bool is_hidden() const;
+	void set_positional(bool p_positional);
+	bool is_positional() const;
+	void set_required(bool p_required);
+	bool is_required() const;
+	void set_multitoken(bool p_multitoken);
+	bool is_multitoken() const;
+
+	void set_static_checker(CheckFunction p_function, const String &p_error_msg);
+	Error set_checker(const Callable &p_callable, const String &p_error_msg);
+	const ArgumentChecker *get_checker() const;
+	void remove_checker();
+
+	void set_data(const String &p_category, const String &p_description);
+
+	CommandLineOption() = default;
+	explicit CommandLineOption(const PackedStringArray &p_names, int p_args_count = 1);
+	~CommandLineOption() override;
+};
+
+class CommandLineParser : public Reference {
+	GDCLASS(CommandLineParser, Reference);
+
+	struct ParsedPrefix;
+
+	inline static const String ARGS_SEPARATOR = "--";
+
+	Vector<Ref<CommandLineOption>> _options;
+
+	PackedStringArray _args;
+	PackedStringArray _forwarded_args;
+
+	PackedStringArray _long_prefixes = sarray("--");
+	PackedStringArray _short_prefixes = sarray("-");
+
+	Map<const CommandLineOption *, PackedStringArray> _parsed_values;
+	Map<const CommandLineOption *, PackedStringArray> _parsed_prefixes;
+	Map<const CommandLineOption *, int> _parsed_count;
+
+	String _help_header;
+	String _help_footer;
+	String _usage_title;
+
+	String _error;
+
+	float _min_option_simmilarity = 0.3;
+
+	int _left_help_pad = 2;
+	int _right_help_pad = 4;
+	int _help_line_length = 80;
+	int _min_description_length = _help_line_length / 2;
+
+	bool _autogenerate_usage = true;
+	bool _allow_forwarding_args = false;
+	bool _allow_adjacent = true;
+	bool _allow_sticky = true;
+	bool _allow_compound = true;
```

Comment:
```suggestion
	bool _allow_forwarding = false;
	bool _allow_adjacent = true;
	bool _allow_sticky = true;
	bool _allow_compound = true;
```

After looking at this again, would this make it consistent or not?

Reply:
Hmm... `_allow_forwarding_args` sounds more explicit to me then `_allow_forwarding`.

Reply:
Perhaps it should not be a property then. Maybe it could be used as an optional boolean parameter in `parse_args(allow_forwarding = false)`? Because it would directly relate to *args* parsing. 🙂

Sorry if I sound picky but that the process I'm going though myself as well while developing my own features, until I feel satisfied/confident in the API. 😄

Reply:
> Maybe it could be used as an optional boolean parameter in parse_args(allow_forwarding = false)

Having bool properties in functions is usually a bad idea. It reduce readability. I think such things should be used only as last resort.
In our case properties for such things sounds fine to me. We have a parser and this is parsing options. This is how it usually done in other parsers.

> Sorry if I sound picky but that the process I'm going though myself as well while developing my own features, until I feel satisfied/confident in the API. smile

It's okay, I'm exactly the same :)